### PR TITLE
Fix memory leak in checkpointer process.

### PR DESCRIPTION
**Pull request by gfphoenix78**
Checkpointer process is a long-live and the current memory context for
the FOR loop is its own memory context. So any memory leak will lead to
the checkpointer process hold more and more memory until the memory
context is reset.
The `rdata` to build xlog record has 5 pointers that allocated by
palloc/palloc0 and only one pointer frees the memory. It's better to
free memory here rather than resetting the memory context in the for loop
of the checkpointer process.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by d** [comment]
Good catch! Do we have a similar leak in 6X_STABLE and / or master?

**Comment by gfphoenix78** [comment]
> Good catch! Do we have a similar leak in 6X_STABLE and / or master?

No, I'll scan the code for master and 6X_STABLE.

**Comment by paul-guo-** [comment]
Off the topic, if gp_before_filespace_setup is set, there is no use of rdata[2-4], but we use rdata[5] for 2pc data, is that another bug?

```
mmxlog_append_checkpoint_data()
{
      rdata[1].next = NULL;
      rdata[2].next = NULL;
      rdata[3].next = NULL;
      rdata[4].next = NULL;

      if (gp_before_filespace_setup)
      {
          if (Debug_persistent_recovery_print)
          {
              SUPPRESS_ERRCONTEXT_DECLARE;

              SUPPRESS_ERRCONTEXT_PUSH();

              elog(PersistentRecovery_DebugPrintLevel(),
                   "mmxlog_append_checkpoint_data: no tablespace and filespace information for checkpoint because gp_before_filespace_setup GUC is true");

              SUPPRESS_ERRCONTEXT_POP();
          }
          return;
      }
}

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -9079,30 +9079,55 @@ CreateCheckPoint(int flags)
...
-	rdata[5].buffer = InvalidBuffer;
-	rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
-	rdata[4].next = &(rdata[5]);
-	rdata[5].next = NULL;
+	if (gp_before_filespace_setup)
+	{
+		rdata[5].data = (char*)p;
+		rdata[5].buffer = InvalidBuffer;
+		rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
+		rdata[1].next = &(rdata[2]);
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c]: This seems wrong, should be `rdata[4].next = &(rdata[5]);`

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -9079,30 +9079,55 @@ CreateCheckPoint(int flags)
 
 	getTwoPhasePreparedTransactionData(&p, "CreateCheckPoint");
 	elog(PersistentRecovery_DebugPrintLevel(), "CreateCheckPoint: prepared transactions = %d", p->count);
-	rdata[5].data = (char*)p;
-	rdata[5].buffer = InvalidBuffer;
-	rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
-	rdata[4].next = &(rdata[5]);
-	rdata[5].next = NULL;
+	if (gp_before_filespace_setup)
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c]: Isn't this condition reversed? gp_before_filespace_setup true means `mmxlog_append_checkpoint_data()` didn't use the `rdata[2...4]`. With current code we are overwriting the content of rdata[2] and losing what `mmxlog_append_checkpoint_data()` had written.

File: src/backend/access/transam/xlog.c [hunk 3]
```diff
@@ -9079,30 +9079,55 @@ CreateCheckPoint(int flags)
 
 	getTwoPhasePreparedTransactionData(&p, "CreateCheckPoint");
 	elog(PersistentRecovery_DebugPrintLevel(), "CreateCheckPoint: prepared transactions = %d", p->count);
-	rdata[5].data = (char*)p;
-	rdata[5].buffer = InvalidBuffer;
-	rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
-	rdata[4].next = &(rdata[5]);
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c]: I feel it's better to just keep this old code and only modify this line, means linking `&(rdata[5]);` to either `rdata[4].next` or `rdata[1].next`. Best would be make that decision my actually following the chain starting rdata[1].next and if null make the chain. So, avoid making decision based on `gp_before_filespace_setup`. Make it purely based on rdata state itself.

Comment (ashwinstar) [reply src/backend/access/transam/xlog.c]: Based on how we decide to modify the code will have to or not modify call to `getTwoPhaseOldestPreparedTransactionXLogRecPtr()` below. With current code modifications its wrong. Mostly likely should be crashing.

File: src/backend/access/transam/xlog.c [hunk 4]
```diff
@@ -9143,7 +9168,24 @@ CreateCheckPoint(int flags)
 			 XLogLastInsertDataLen());
 	}
 
+	/* free heap memory allocated by palloc, to prevent memory leak in the long-live memory context */
+	/* rdata[0].data uses the stack buffer */
+	/* rdata[1] is dtxCheckPointInfo */
 	freeDtxCheckPointInfoAndUnlock(dtxCheckPointInfo, dtxCheckPointInfoSize, &recptr);
+	if (gp_before_filespace_setup)
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c]: I strongly feel we should not decide based on this guc `gp_before_filespace_setup` to pfree or not. Better write the loop following rdata chain and if not null `pfree()` the same.

File: src/backend/access/transam/xlog.c (hunk 2 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Ah, yes. I'm wrong.
> With current code we are overwriting the content of rdata[2] and losing what mmxlog_append_checkpoint_data() had written.

Good idea.

File: src/backend/access/transam/xlog.c (hunk 4 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: I'd rather not. rdata[0].data points to the stack buffer, and dtxCheckPointInfo has its own free function.
Explicitly calling `pfree()` makes the logic more clear.

File: src/backend/access/transam/xlog.c (hunk 1 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Yes, many thanks.

File: src/backend/access/transam/xlog.c (hunk 4 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: I thought again and updated the code as you suggested with some comments to tell people where the rdata comes from.

File: src/backend/access/transam/xlog.c [hunk 5]
```diff
@@ -9143,7 +9147,22 @@ CreateCheckPoint(int flags)
 			 XLogLastInsertDataLen());
 	}
 
+	/* free heap memory allocated by palloc, to prevent memory leak in the long-live memory context */
+	/* rdata[0].data uses the stack buffer */
+	/* rdata[1] is dtxCheckPointInfo */
 	freeDtxCheckPointInfoAndUnlock(dtxCheckPointInfo, dtxCheckPointInfoSize, &recptr);
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c]: I think we should just modify `freeDtxCheckPointInfoAndUnlock` to only release lock and avoid passing `dtxCheckPointInfo` to it. Passing the pointer just to call `pfree()` seems dumb. Remove the local variable `dtxCheckPointInfo` altogether. `getDtxCheckPointInfoAndLock()` can be passed rdata[5] itself.

This way can `pfree()` all the `rdata[].data` in one place here.

File: src/backend/access/transam/xlog.c [hunk 6]
```diff
@@ -9143,7 +9147,22 @@ CreateCheckPoint(int flags)
 			 XLogLastInsertDataLen());
 	}
 
+	/* free heap memory allocated by palloc, to prevent memory leak in the long-live memory context */
```
[the remark this answers is not in the document]
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Thanks, changed

File: src/backend/access/transam/xlog.c [hunk 7]
```diff
@@ -9082,8 +9082,12 @@ CreateCheckPoint(int flags)
 	rdata[5].data = (char*)p;
 	rdata[5].buffer = InvalidBuffer;
 	rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
-	rdata[4].next = &(rdata[5]);
 	rdata[5].next = NULL;
+	/* if !gp_before_filespace_setup then rdata[2...4] are valid */
+	if (gp_before_filespace_setup)
```
[the remark this answers is not in the document]
Comment (ashwinstar) [reply src/backend/access/transam/xlog.c:9087]: I am not seeing review comments being incorporated in the follow-up changes? Please notify me when the things are fixed and PR ready to be reviewed again. In current state I don't feel comfortable committing this change.

[the remark this answers is not in the document]
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Thank you for your patience. There are some test failures I'm trying to fix. I'll let you know as soon as it's ready.

**Comment by gfphoenix78** [comment]
rdata[x].data may be changed in the function `XLogInsert()`.
So we can't call `pfree(rdata[xxx].data);`

File: src/backend/access/transam/xlog.c (hunk 5 above)
Comment (ashwinstar) [reply src/backend/access/transam/xlog.c:9173]: This comment is not addressed yet.

File: src/backend/access/transam/xlog_mm.c [hunk 8]
```diff
@@ -1598,6 +1598,13 @@ mmxlog_append_checkpoint_data(XLogRecData rdata[6])
 				 "mmxlog_append_checkpoint_data: no tablespace and filespace information for checkpoint because gp_before_filespace_setup GUC is true");
 
 			SUPPRESS_ERRCONTEXT_POP();
+
+			rdata[2].data = NULL;
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog_mm.c:1602]: Is there a better place to make these initialization, maybe in CreateCheckPoint() itself?

File: src/backend/access/transam/xlog.c [hunk 9]
```diff
@@ -9082,8 +9083,28 @@ CreateCheckPoint(int flags)
...
 	rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
-	rdata[4].next = &(rdata[5]);
 	rdata[5].next = NULL;
+	/*
+	 * We iterate the rdata because rdata[2,3,4] may or may not be used
+	 * in mmxlog_append_checkpoint_data(). The next pointers of the items
+	 * 2 ~ 4(inclusive) are safely initialized to NULL in
+	 * mmxlog_append_checkpoint_data().
+	 */
+	for (i = 1; i < 5; i++)
```
[the remark this answers is not in the document]
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Updated.

File: src/backend/access/transam/xlog_mm.c (hunk 8 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog_mm.c]: The function mmxlog_append_checkpoint_data() itself is too large to put it in CreateCheckPoint(). `mmxlog_append_checkpoint_data()` may output 3 XLogRecData, I let this function to take care of the link-list.

File: src/backend/access/transam/xlog.c [hunk 10]
```diff
@@ -9055,35 +9053,50 @@ CreateCheckPoint(int flags)
...
+	XLogRecData **pnext;
 
 	rdata[0].data = (char *) (&checkPoint);
 	rdata[0].len = sizeof(checkPoint);
 	rdata[0].buffer = InvalidBuffer;
 	rdata[0].next = &(rdata[1]);
 
-	rdata[1].data = (char *) dtxCheckPointInfo;
-	rdata[1].len = dtxCheckPointInfoSize;
+	getDtxCheckPointInfoAndLock((char **)&rdata[1].data, (int *)&rdata[1].len);
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:9064]: I don't know why we changed this piece of code. Previous code was fine and easier to read. If you wish to restrict the usage of variable can easily put them in its scope like

```
{
    char* 		dtxCheckPointInfo;
    int			dtxCheckPointInfoSize;

    rdata[1].data = (char *) dtxCheckPointInfo;
    rdata[1].len = dtxCheckPointInfoSize;
}
```
I am not fan of the way code looks with change, its unnecessarily complicated.

File: src/backend/access/transam/xlog.c [hunk 11]
```diff
@@ -9055,35 +9053,50 @@ CreateCheckPoint(int flags)
...
 	rdata[1].buffer = InvalidBuffer;
-	rdata[1].next = NULL;
+	pnext = &rdata[1].next;
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:9066]: I would still like to keep the `rdata[1].next = NULL;` part. It's okay it gets overwritten with valid value of NULL again later.

File: src/backend/access/transam/xlog.c [hunk 12]
```diff
@@ -9055,35 +9053,50 @@ CreateCheckPoint(int flags)
 	 * dropped on the master, break the consistency between the master and the standby.
 	 */
 
-	getDtxCheckPointInfoAndLock(&dtxCheckPointInfo, &dtxCheckPointInfoSize);
+	/* pnext is used to link the valid XLogRecData */
+	XLogRecData **pnext;
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:9057]: define this at start of function

File: src/backend/access/transam/xlog.c [hunk 13]
```diff
@@ -9055,35 +9053,50 @@ CreateCheckPoint(int flags)
...
 	rdata[5].data = (char*)p;
 	rdata[5].buffer = InvalidBuffer;
 	rdata[5].len = PREPARED_TRANSACTION_CHECKPOINT_BYTES(p->count);
-	rdata[4].next = &(rdata[5]);
 	rdata[5].next = NULL;
+	/*
+	 * Save the data pointers that need to be pfreed after XLogInsert.
+	 * This is essential because rdata[x].data may be changed in XLogInsert.
+	 */
+	void *dataptr[6];
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:9090]: define this at start of function

File: src/backend/access/transam/xlog.c [hunk 14]
```diff
@@ -9143,7 +9156,16 @@ CreateCheckPoint(int flags)
 			 XLogLastInsertDataLen());
 	}
 
-	freeDtxCheckPointInfoAndUnlock(dtxCheckPointInfo, dtxCheckPointInfoSize, &recptr);
+	freeDtxCheckPointInfoAndUnlock(&recptr);
+	/*
+	 * if gp_before_filespace_setup is on, rdata[2], rdata[3], rdata[4] are empty,
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:9161]: No need for this comment

File: src/backend/access/transam/xlog.c (hunk 10 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Thanks.

File: src/backend/access/transam/xlog.c (hunk 11 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c:9073]: Changed.

File: src/backend/access/transam/xlog.c (hunk 12 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Good point.

File: src/backend/access/transam/xlog.c (hunk 13 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Good point.

File: src/backend/access/transam/xlog.c (hunk 14 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Removed.

**Pull request merged**